app/custom-oauth/server/oauth_helpers.js

```diff
@@ -0,0 +1,33 @@
+
+import { Roles } from '../../models';
+
+/**
+*/
+export function mapRolesFromSSO(user, identity) {
+	// Adding roles
+	if (user && identity.roles) {
+		// loop through all assigned roles and drop if not assigned anymore
+		user.roles.forEach(function(roleFromOAuthProvider) {
+			if (identity.roles.length === 0) {
+				Roles.removeUserFromRoles(user._id, roleFromOAuthProvider);
+			}
+			if (identity.roles.indexOf(roleFromOAuthProvider) === -1) {
+				Roles.removeUserFromRoles(user._id, roleFromOAuthProvider);
+			}
+		});
+		// loop through all roles from SSO Provider and add them if needed
+		identity.roles.forEach(function(roleFromOAuthProvider) {
+			if (user.roles.indexOf(roleFromOAuthProvider) === -1
+				/* filter OpenID pseudo roles */
+				&& roleFromOAuthProvider !== 'offline_access' && roleFromOAuthProvider !== 'uma_authorization') {
```

Comment:
I got to wonder.. if these are present in roles.. isn't this essentially scopes?

Reply:
also do we need to set the property roles is pulled from?  I personally haven't seen an oauth provider that delivers roles.  So would be good to see some examples?  Is this specific property part of the spec?

Reply:
you're right it's a claim that needs to be defined. E.g. normally groups is the claim used by many OAuth server implementations, e.g. keycloak.
I will add a settings param for this